Prompt user to enable EFA for supported instance-types when using the pcluster configure wizard

EddyMM opened the pull request:
### Description of changes
* Users who create a cluster configuration using `pcluster configure` have minimal awareness of EFA being able to optimise performance for supported instance types
* These changes aim to address this by taking the following actions when a user specifies an EFA-supported compute instance type in `pcluster configure`:
   * Explain the benefits of enabling EFA
   * Prompt the user to enable EFA (or choose not to)
   * If the user enables EFA, they will be prompted to either provide an existing Placement Group or let parallelCluster create one for them

### Tests
* Tests have been added for the EFA prompt workflow to confirm the appropriate configuration is generated
  * If EFA is enabled, and existing placement group is provided the appropriate output and configuration file is generated
  * If EFA is enabled, and non-existent placement group is provided, the user is informed that the Placement Group is unknown and requested to try again
  * If EFA is enabled, and no placement group is provided (ParallelCluster will create one by default) the appropriate output and configuration file is generated
  * If EFA is NOT enabled, the user is not prompted to provide a Placement Group
  * If EFA is NOT enabled, the appropriate output and configuration file is generated

### References
* [ParallelCluster EFA](https://docs.aws.amazon.com/parallelcluster/latest/ug/efa.html)
* [AWS EFA](https://aws.amazon.com/hpc/efa/)

### Checklist
- [ ] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

enrico-usai commented on CHANGELOG.md:
```diff
@@ -14,6 +14,7 @@ x.x.x
 - Show `requested_value` and `current_value` values in the change set when adding or removing a section.
 - Add new configuration parameter `Scheduling/SlurmSettings/EnableMemoryBasedScheduling` to configure memory-based
   scheduling in Slurm.
+- Prompt user to enable EFA for supported instance types when using `pcluster configure` wizard
```
minor: missing final period.

EddyMM replied:
Done.

enrico-usai commented on cli/src/pcluster/cli/commands/configure/easyconfig.py:
```diff
@@ -232,25 +243,47 @@ def configure(args):  # noqa: C901
+                    "MaxCount": max_cluster_size,
+                }
+                if efa_supported_by_instance_type:
+                    if efa_enabled_in_compute_resource:
```
minor: you can rewrite the entire if/else block to:
```python
compute_resource["Efa"] = {"Enabled": efa_enabled_in_compute_resource}
```

EddyMM replied:
Done.

enrico-usai commented on cli/src/pcluster/cli/commands/configure/easyconfig.py:
```diff
@@ -232,25 +243,47 @@ def configure(args):  # noqa: C901
+        placement_group = {}
+        if efa_enabled_in_queue:
+            placement_group_name = _prompt_for_placement_group()
+            if placement_group_name:
```
minor: to avoid code duplication you can write:
```python
placement_group = {"Enabled": True}
if placement_group_name:
    placement_group["Id"] = placement_group_name

```

EddyMM replied:
Done.

enrico-usai commented on cli/src/pcluster/cli/commands/configure/easyconfig.py:
```diff
@@ -232,25 +243,47 @@ def configure(args):  # noqa: C901
     # Here is the end of prompt. Code below assembles config and write to file
     for queue in queues:
-        queue["Networking"] = {"SubnetIds": [vpc_parameters["compute_subnet_id"]]}
+        if "Networking" in queue:
```
an alternative to avoid if/else is:
```python
networking = queue.get("Networking", {})
networking["SubnetIds"] = [vpc_parameters["compute_subnet_id"]]
queue["Networking"] = networking
```

EddyMM replied:
Done.

enrico-usai commented on cli/src/pcluster/cli/commands/configure/easyconfig.py:
```diff
@@ -232,25 +243,47 @@ def configure(args):  # noqa: C901
+            "ComputeResources": compute_resources,
+        }
+        if placement_group:
+            queue.update({"Networking": {"PlacementGroup": placement_group}})
```
If networking contains other fields they will be removed, we should use the same logic used below

enrico-usai commented on cli/src/pcluster/cli/commands/configure/utils.py:
```diff
@@ -203,6 +205,32 @@ def get_regions():
+
+    It does so by querying the AWS EC2 API for the placement group by name.
+    :param placement_group_name: Name of the placement group
+    :return: True if placement group exists, False if placement group does not exist or unknown ina  region
```
minor: typo at the end

enrico-usai commented on cli/src/pcluster/cli/commands/configure/easyconfig.py:
```diff
@@ -232,25 +243,47 @@ def configure(args):  # noqa: C901
+            "Name": queue_name,
+            "ComputeResources": compute_resources,
+        }
+        if placement_group:
```
you could move the initialization of the queue dict before the `if efa_enabled_in_queue` block, then move this `if placement_group` into the `if efa_enabled_in_queue` block.

EddyMM replied:
Done.

demartinofra commented on cli/src/pcluster/cli/commands/configure/utils.py:
```diff
@@ -203,6 +205,32 @@ def get_regions():
     return regions
 
 
+@handle_client_exception
```
`AWSApi.instance().ec2.get_instance_type_info` is already decorated with `@AWSExceptionHandler.handle_client_exception` which translates botocore exceptions in internal exceptions. I'm wondering if the handle_client_exception here works as you expect it to.

EddyMM replied:
`@AWSExceptionHandler.handle_client_exception` will ultimately raise an error.
`@handle_client_exception` from `utils` will only catch the exception and print out the Client Error, which is the behaviour we need here.

demartinofra replied:
But are you sure the correct exception is catched by handle_client_exception from utils? Cause I see a mismatch between what @AWSExceptionHandler.handle_client_exception raises and what the other one handles

EddyMM replied:
`handle_client_exception` from utils catches the AWSClientError exception raised from `@AWSExceptionHandler.handle_client_exception`. Only difference is that the utils version will handle it by printing the exception (instead of raising it the next level in the hierarchy & potentially ultimately exiting).

demartinofra replied:
Oh ok sorry for some reason I had missed your change where you added AWSClientError to the except cause

demartinofra commented on cli/src/pcluster/cli/commands/configure/utils.py:
```diff
@@ -203,6 +205,32 @@ def get_regions():
+    return AWSApi.instance().ec2.get_instance_type_info(instance_type).is_efa_supported()
+
+
+@handle_client_exception
```
same here

enrico-usai commented on cli/src/pcluster/cli/commands/configure/utils.py:
```diff
@@ -203,6 +205,32 @@ def get_regions():
+    try:
+        return AWSApi.instance().ec2.describe_placement_group(placement_group_name)
+    except AWSClientError as e:
+        print(f"Failed with error: {e}")
```
I don't think we want to print the error here. Should be in charge of @handle_client_exception

demartinofra commented on cli/src/pcluster/cli/commands/configure/easyconfig.py:
```diff
@@ -232,25 +243,47 @@ def configure(args):  # noqa: C901
+                    if efa_enabled_in_compute_resource:
+                        compute_resource["Efa"] = {"Enabled": True}
+                    else:
+                        compute_resource["Efa"] = {"Enabled": False}
```
nit: it's a matter of personal taste but I find more readable a compact form like:

```
compute_resource = {
                    "Name": compute_resource_name,
                    "InstanceType": compute_instance_type,
                    "MinCount": min_cluster_size,
                    "MaxCount": max_cluster_size,
                    "Efa": { "Enabled": efa_supported_by_instance_type and efa_enabled_in_compute_resource }
                }
```

EddyMM replied:
Yes, I'll keep the outer `if` though, to avoid defining the "Efa" property for instance-types that don't have support for it anyway.

```
if efa_supported_by_instance_type:
    compute_resource["Efa"] = {"Enabled": efa_enabled_in_compute_resource}
```

demartinofra commented on cli/src/pcluster/cli/commands/configure/easyconfig.py:
```diff
@@ -232,25 +243,47 @@ def configure(args):  # noqa: C901
             cluster_size += max_cluster_size  # Fixme: is it the right calculation for awsbatch?
-        queues.append({"Name": queue_name, "ComputeResources": compute_resources})
+
+        placement_group = {}
```
Dropping it here but it's more of a general comment: this configure function is getting extremely long and hard to parse. Not asking to refactor the entire code as part of this PR cause the effort might be too big but if you could come up with some partial improvements especially around the areas you are modifying it would be beneficial for code quality

EddyMM replied:
I agree. At the moment the `configure` function is not properly modularised. I tried to move the prompting logic to separate functions but a refactor may be needed later for other prompts and logic.

demartinofra commented on cli/src/pcluster/cli/commands/configure/easyconfig.py:
```diff
@@ -401,9 +434,27 @@ def _prompt_for_subnet(all_subnets, qualified_subnets, message, default_subnet=N
 
 
+def _prompt_for_placement_group():
+    print(
+        "Enabling EFA requires compute instances to be placed within a Placement Group, Specify an existing "
+        "Placement Group name or leave blank for ParallelCluster to create one"
+    )
```
What if I don't want a PG? shall we allow Y/N or PG name selection?

EddyMM replied:
- With the wizard workflow, we went for requiring a placement group with EFA.
- The placement group can always be set to `false` in the config file but for the initial configuration we're trying to get the most efficient setup

demartinofra commented on cli/tests/pcluster/cli/configure/test_pcluster_configure.py:
```diff
@@ -689,14 +692,88 @@ def general_wrapper_for_prompt_testing(
     return True
 
 
+def test_enabled_efa_default_placement_group(mocker, temp_path_for_config, capsys, test_datadir):
```
nit: if the newly added tests share most of the test code you could consider replacing them with a single parametrized test_efa test case where pg setting is a parameter

demartinofra commented on cli/tests/pcluster/cli/configure/test_pcluster_configure/test_enabled_efa_default_placement_group/output.txt:
```diff
@@ -0,0 +1,57 @@
+INFO: Configuration file {{ CONFIG_FILE }} will be written.
```
Don't we capture and validate the questions/prompts as part of the unit testing? is there a technical limitation?

EddyMM replied:
We only capture the output, errors and configs at the moment.

demartinofra replied:
is there an easy way to also capture the prompted question? I think that would be extremely valuable to test with the added value of having the entire configure workflow dumped in a file

EddyMM replied:
I can take a look at the mechanism used to capture the input to see if it's possible.

enrico-usai commented on cli/tests/pcluster/cli/configure/test_pcluster_configure.py:
```diff
@@ -689,14 +692,64 @@ def general_wrapper_for_prompt_testing(
+@pytest.mark.parametrize(
+    "test_dir_name, supports_efa, efa_enabled, placement_group_names, placement_group_exists",
+    [
+        ("test_enabled_efa_default_placement_group", True, "y", [""], lambda x: False),
```
what about a test with supports_efa = False?

EddyMM replied:
Most of the other tests in `test_pcluster_configure.py` cover such a scenario, but I can explicitly include it here to be sure.

EddyMM replied:
Added a scenario where `supports_efa = False`

EddyMM committed:
Prompt user to enable EFA and Placement Group for Supported Instance Types

- To improve awareness and adoption of EFA(https://docs.aws.amazon.com/parallelcluster/latest/ug/efa.html) for instance types that support it, this commit introduces a prompt to inform the user of the benefits and request them to enable it.
- Using a placement group with EFA gives better results for tightly-coupled workloads so if a user enables EFA, we also ask them if they want to add the queue to a placement group or otherwise we create one for them

Signed-off-by: Eddy Mwiti <eddmwiti@amazon.com>
